File: mobile/lib/models/product.dart
import 'package:json_annotation/json_annotation.dart';
import '../core/constants/api_constants.dart';

part 'product.g.dart';

@JsonSerializable()
class ProductVariant {
  @JsonKey(name: '_id')
  final String id;
  final String label;
  final int weightGrams;
  final String sku;
  final double price;
  final double? compareAtPrice;
  final int stockQty;
  final bool active;

  ProductVariant({
    required this.id,
    required this.label,
    required this.weightGrams,
    required this.sku,
    required this.price,
    this.compareAtPrice,
    this.stockQty = 0,
    this.active = true,
  });

  bool get inStock => stockQty > 0 && active;

  factory ProductVariant.fromJson(Map<String, dynamic> json) =>
      _$ProductVariantFromJson(json);
  Map<String, dynamic> toJson() => _$ProductVariantToJson(this);
}

@JsonSerializable()
class ProductImage {
  final String? secureUrl;
  final String? cloudinaryPublicId;

  ProductImage({this.secureUrl, this.cloudinaryPublicId});

  factory ProductImage.fromJson(Map<String, dynamic> json) =>
      _$ProductImageFromJson(json);
  Map<String, dynamic> toJson() => _$ProductImageToJson(this);
}

@JsonSerializable()
class Product {
  @JsonKey(name: '_id')
  final String id;
  final String name;
  final String slug;
  final String? category;
  final String? region;
  final double basePrice;
  final String? roast;
  final String? image;
  final List<String>? features;
  final String? description;
  final String? tastingNotes;
  final String? processing;
  final bool bestseller;
  final bool inStock;
  final bool active;
  final List<ProductImage>? images;
  final List<ProductVariant> variants;

  Product({
    required this.id,
    required this.name,
    required this.slug,
    this.category,
    this.region,
    required this.basePrice,
    this.roast,
    this.image,
    this.features,
    this.description,
    this.tastingNotes,
    this.processing,
    this.bestseller = false,
    this.inStock = true,
    this.active = true,
    this.images,
    this.variants = const [],
  });

  String get displayImage =>
      ApiConstants.getImageUrl(images?.firstOrNull?.secureUrl ?? image);

  ProductVariant? get defaultVariant =>
      variants.where((v) => v.active && v.stockQty > 0).firstOrNull ??
      variants.firstOrNull;

  double get displayPrice => defaultVariant?.price ?? basePrice;

  bool get hasVariantsInStock => variants.any((v) => v.inStock);

  factory Product.fromJson(Map<String, dynamic> json) =>
      _$ProductFromJson(json);
  Map<String, dynamic> toJson() => _$ProductToJson(this);
}

File: backend/models/Product.js
const mongoose = require('mongoose');

const VariantSchema = new mongoose.Schema({
  label: { type: String, required: true },        // "250g", "500g", "1kg"
  weightGrams: { type: Number, required: true },  // 250, 500, 1000
  sku: { type: String, required: true, unique: true },
  price: { type: Number, required: true },        // FINAL price used in checkout
  compareAtPrice: { type: Number },               // optional strike-through price
  stockQty: { type: Number, default: 0 },
  active: { type: Boolean, default: true }
}, { _id: true });

const ProductSchema = new mongoose.Schema({
  name: { type: String, required: true },
  slug: { type: String, unique: true },           // SEO + Flutter friendly

  category: { 
    type: String, 
    required: true,
    enum: ['africa', 'america', 'asia', 'premium', 'all']
  },

  region: { type: String, required: true },

  // Keep for backward compatibility (DO NOT USE FOR CHECKOUT)
  basePrice: { type: Number, required: true },

  roast: { 
    type: String, 
    required: true,
    enum: ['Light', 'Medium', 'Dark']
  },

  // Primary image
  image: { type: String, required: true },

  // Future-ready gallery
  images: [{
    secureUrl: String,
    cloudinaryPublicId: String
  }],

  variants: [VariantSchema],   // ⭐ CORE ECOMMERCE - use for checkout

  features: [{ type: String }],
  description: { type: String, required: true },
  tastingNotes: { type: String },
  processing: { type: String, default: 'Washed' },

  bestseller: { type: Boolean, default: false },
  active: { type: Boolean, default: true },

  // Computed dynamically from variants
  inStock: { type: Boolean, default: true }

}, { timestamps: true });

// Auto-generate slug + compute inStock
ProductSchema.pre('save', function(next) {
  if (!this.slug) {
    this.slug = this.name
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, '-')
      .replace(/(^-|-$)/g, '');
  }

  // Auto-compute inStock from variants
  if (this.variants && this.variants.length) {
    this.inStock = this.variants.some(v => v.active && v.stockQty > 0);
  }

  next();
});

module.exports = mongoose.model('Product', ProductSchema);
